extensions/wikia/AdEngine/js/video/googleIma.js

```diff
@@ -36,43 +36,88 @@ define('ext.wikia.adEngine.video.googleIma', [
 		return videoAdContainer;
 	}
 
-	function setupIma(vastUrl, adContainer, width, height) {
-		isAdsManagerLoaded = false;
-		adDisplayContainer = new google.ima.AdDisplayContainer(adContainer);
-		adsLoader = new google.ima.AdsLoader(adDisplayContainer);
-		adsLoader.addEventListener(google.ima.AdsManagerLoadedEvent.Type.ADS_MANAGER_LOADED, onAdsManagerLoaded, false);
-		adsLoader.requestAds(createRequest(vastUrl, width, height));
+	function registerEvents(ima) {
+		Object.keys(google.ima.AdEvent.Type).forEach(function (eventKey) {
+			var eventName = google.ima.AdEvent.Type[eventKey];
+			ima.adsManager.addEventListener(eventName, function (event) {
+				ima.events[eventName] = ima.events[eventName] || [];
 
-		adDisplayContainer.initialize();
+				ima.events[eventName].map(function (callback) {
+					callback(ima, event);
+				});
+				log([eventName, event.getAdData()], log.levels.debug, logGroup);
+			}, false);
+		});
 
-		return prepareVideoAdContainer(adContainer.querySelector('div'));
-	}
+		Object.keys(google.ima.AdErrorEvent.Type).forEach(function (eventKey) {
```

Comment:
I see kinda similar code for `google.ima.AdErrorEvent.Type` and `google.ima.AdEvent.Type`, didn't you think about abstracting it?